### PR TITLE
[PIL-634][OK] - Added logic to catch 409 for duplicate acknowledgement

### PR DESCRIPTION
**Pull request by olakunlekuye2**

### CODE REVIEW

File: app/controllers/AlreadyRegisteredController.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+class AlreadyRegisteredController @Inject() (val controllerComponents: MessagesControllerComponents, view: AlreadyRegisteredView)(implicit
+  ec:                                                                  ExecutionContext,
+  appConfig:                                                           FrontendAppConfig
+) extends FrontendBaseController
+    with I18nSupport {
+
+  def onPageLoad: Action[AnyContent] = Action { implicit request =>
+    Ok(view())
+  }
+}
```
Comment (jassalrichy) [inline app/controllers/AlreadyRegisteredController.scala:37]: Can we add a spec for this controller.

File: test/connectors/EnrolmentStoreProxyConnectorSpec.scala [hunk 2]
```diff
@@ -57,22 +57,22 @@ class EnrolmentStoreProxyConnectorSpec extends SpecBase {
...
-      result.futureValue mustBe false
+      result.futureValue mustBe Right(false)
     }
 
     "return 204 enrolmentStatus response when principalGroupId is empty seq" in {
       val plrRef = "xxx204"
       stubGet(enrolmentStoreProxy204Url, OK, enrolmentStoreProxyResponseNoPrincipalIdJson)
       val result = connector.enrolmentExists(plrRef)
-      result.futureValue mustBe false
+      result.futureValue mustBe Right(false)
```
Comment (jassalrichy) [inline test/connectors/EnrolmentStoreProxyConnectorSpec.scala:75]: Do we need to add a test for a 409 status return ?

Comment (olakunlekuye2) [reply test/connectors/EnrolmentStoreProxyConnectorSpec.scala:75]: Done

File: app/controllers/AlreadyRegisteredController.scala (hunk 1 above)
Comment (olakunlekuye2) [reply app/controllers/AlreadyRegisteredController.scala:37]: Done

File: app/connectors/SubscriptionConnector.scala [hunk 3]
```diff
@@ -35,22 +39,66 @@ class SubscriptionConnector @Inject() (val userAnswersConnectors: UserAnswersCon
...
+//          None
+//      }
+//      .recover { case e: Exception =>
+//        logger.warn(
+//          s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - " +
+//            s"Error message ${e.printStackTrace()} has been thrown when create subscription was called"
+//        )
+//        None
+//      }
+
```
Comment (jassalrichy) [inline app/connectors/SubscriptionConnector.scala:103]: Remove commented out code.

File: app/views/errors/AlreadyRegisteredView.scala.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+@this(
+layout: templates.Layout
+)
+
+
+@()(implicit request: Request[_],  appConfig: FrontendAppConfig, messages: Messages)
+
+@layout(pageTitle = titleNoForm(messages("alreadyRegistered.title")), showBackLink = false) {
+<h1 class="govuk-heading-xl">@messages("alreadyRegistered.heading")</h1>
+val email = "pillar2mailbox.gov.uk"
```
Comment (jassalrichy) [inline app/views/errors/AlreadyRegisteredView.scala.html:27]: Remove this line.

File: app/views/errors/AlreadyRegisteredView.scala.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+<p class="govuk-body">@messages("alreadyRegistered.message1")</p>
+<p class="govuk-body">
+    @Html(messages("alreadyRegistered.message2", s"<a class='govuk-link' href='mailto:${appConfig.contactEmail}'>${appConfig.contactEmail}</a>"))
+</p>
```
Comment (jassalrichy) [inline app/views/errors/AlreadyRegisteredView.scala.html:31]: email address different from that on Jira ticket.

File: app/connectors/SubscriptionConnector.scala [hunk 6]
```diff
@@ -35,22 +39,66 @@ class SubscriptionConnector @Inject() (val userAnswersConnectors: UserAnswersCon
...
+                logger.info("Successfully deserialized response")
+                Some(successResponse.success)
+              case _ =>
+                logger.error("Failed to deserialize success response")
+                None
+            }
+
+          case CONFLICT =>
+            logger.info(s"Subscription call failed due to conflict with status $CONFLICT and body: ${response.body}")
+            Some(SubscriptionResponse("", "", LocalDateTime.now(), Some(ErrorResponse("Conflict detected during subscription"))))
```
Comment (AKKUM) [inline app/connectors/SubscriptionConnector.scala:59]: I think this is not right way to construct Subscription response. SubscriptionResponse comes from Hod and populates. But here we have mocked and passed it controller. Better to pass the natural HTTPResponse and deal with status and objects on controller or service layers.

File: app/connectors/SubscriptionConnector.scala [hunk 7]
```diff
@@ -35,22 +39,66 @@ class SubscriptionConnector @Inject() (val userAnswersConnectors: UserAnswersCon
...
-      .recover { case e: Exception =>
-        logger.warn(
-          s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - " +
-            s"Error message ${e.printStackTrace()} has been thrown when create subscription was called"
-        )
-        None
+      .recover {
+
+        case UpstreamErrorResponse(message, CONFLICT, _, _) =>
+          logger.info(s"Conflict detected with body: $message")
```
Comment (AKKUM) [inline app/connectors/SubscriptionConnector.scala:69]: Error should go as logger. warn or error.

File: app/connectors/SubscriptionConnector.scala (hunk 3 above)
Comment (olakunlekuye2) [reply app/connectors/SubscriptionConnector.scala:103]: Done

File: app/views/errors/AlreadyRegisteredView.scala.html (hunk 4 above)
Comment (olakunlekuye2) [reply app/views/errors/AlreadyRegisteredView.scala.html:27]: Done

File: app/connectors/SubscriptionConnector.scala (hunk 6 above)
Comment (olakunlekuye2) [reply app/connectors/SubscriptionConnector.scala:59]: I will remove this as it is redundant

File: app/connectors/SubscriptionConnector.scala (hunk 7 above)
Comment (olakunlekuye2) [reply app/connectors/SubscriptionConnector.scala:69]: Okay

File: app/views/errors/AlreadyRegisteredView.scala.html (hunk 5 above)
Comment (olakunlekuye2) [reply app/views/errors/AlreadyRegisteredView.scala.html:31]: Fixed

File: app/views/errors/AlreadyRegisteredView.scala.html [hunk 8]
```diff
@@ -27,7 +27,7 @@ <h1 class="govuk-heading-xl">@messages("alreadyRegistered.heading")</h1>
 val email = "pillar2mailbox.gov.uk"
 <p class="govuk-body">@messages("alreadyRegistered.message1")</p>
 <p class="govuk-body">
-    @Html(messages("alreadyRegistered.message2", s"<a class='govuk-link' href='mailto:${appConfig.contactEmail}'>${appConfig.contactEmail}</a>"))
+    @Html(messages("alreadyRegistered.message2", s"<a class='govuk-link' href='mailto:${appConfig.contactEmail}'>${appConfig.pillar2mailbox}</a>"))
 </p>
```
Comment (jassalrichy) [inline app/views/errors/AlreadyRegisteredView.scala.html:31]: appConfig.contactEmail needs to be updated also ?

Comment (olakunlekuye2) [reply app/views/errors/AlreadyRegisteredView.scala.html:31]: updated to what

File: app/connectors/SubscriptionConnector.scala [hunk 9]
```diff
@@ -35,22 +36,34 @@ class SubscriptionConnector @Inject() (val userAnswersConnectors: UserAnswersCon
...
-        logger.warn(
-          s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - " +
-            s"Error message ${e.printStackTrace()} has been thrown when create subscription was called"
-        )
-        None
+      .recover {
+
+        case UpstreamErrorResponse(message, CONFLICT, _, _) =>
+          logger.error(s"Conflict detected with body: $message")
+          Some(SubscriptionResponse("", "", LocalDateTime.now(), Some(ErrorResponse("Conflict detected during subscription"))))
```
Comment (AKKUM) [inline app/connectors/SubscriptionConnector.scala:62]: Again this cannot be mocked here as well. Please return HttpResponse

File: conf/messages.en [hunk 10]
```diff
@@ -52,7 +52,11 @@ cannotReturnAfterSubscriptionError.message= You have successfully registered to
 cannotReturnAfterSubscriptionError.message2= You can now
 cannotReturnAfterSubscriptionError.link= report and manage your pillar 2 top-up taxes.
 
-
+alreadyRegistered.title= Your group is already registered
+alreadyRegistered.heading= Your group is already registered
+alreadyRegistered.message1= You will need to contact your administrator to give you access.
+alreadyRegistered.message2= You can email {0} if you are having problems accessing the service.
+alreadyRegistered.newtab = Is this page not working properly? (opens in new tab)
```
Comment (jassalrichy) [inline conf/messages.en:59]: alreadyRegistered.newtab can be removed.

File: test/controllers/AlreadyRegisteredControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    "must include specific messages or links" in {
+      val application = applicationBuilder(userAnswers = None).build()
+
+      running(application) {
+        val request = FakeRequest(GET, routes.AlreadyRegisteredController.onPageLoad.url)
+        val result  = route(application, request).value
+
+        val content = contentAsString(result)
+        content must include(messages(application)("alreadyRegistered.message1"))
+        content must include(messages(application)("alreadyRegistered.newtab"))
```
Comment (jassalrichy) [inline test/controllers/AlreadyRegisteredControllerSpec.scala:72]: remove:
content must include(messages(application)("alreadyRegistered.newtab"))

**Review by jassalrichy — APPROVED** [review]

**Review by AKKUM — APPROVED** [review]

**Review by jassalrichy — APPROVED** [review]

**Review by javed-iqbal1 — APPROVED** [review]

**Pull request merged**